### ISSUE TITLE
Pure HTML templates like VueJS `.vue` files

### ISSUE THREAD

**Issue by ThibaudDauce** [slamdata/purescript-halogen#553]
Hi,

Could I use raw HTML with Halogen? VueJS is using [HTML-based template syntax](https://vuejs.org/v2/guide/syntax.html) instead of render functions, and these templates are compiled (at compilation time) to render functions.

I think the HTML DSL of Halogen (or Elm…) is very ugly and I would rather use pure HTML with custom attributes and brackets. To my knowledge, this solution doesn't prevent compile-time checks, but I may be wrong :-) Do you know why Halogen (and other frameworks) chose to use a custom syntax instead of HTML? Is it possible to implement this kind of feature in Halogen?

**Comment by garyb** [comment]
> Is it possible to implement this kind of feature in Halogen?

It's not something we'd implement as part of the library, no. It's technically possible, it would just be a separate program that compiles templates into PureScript code though.

> Do you know why Halogen (and other frameworks) chose to use a custom syntax instead of HTML?

I would say first of all, because it's much easier! But even after the initial hurdle of building a template compiler, there are some other considerations:

- It's only ugly because it's less familiar, it's just as readable after a short period of working with it
- In Halogen's case, it's not really "custom syntax" - you're directly building the virtual DOM tree, so even if there was a templating language it would be translated into this anyway
- A HTML templating language is always going to be less expressive than just writing PS code
- The editor tooling for PS can help you out, `?hole`s etc. still work, there's immediate feedback if you try and use a property that doesn't exist for an element, etc. I guess you could write editor tooling for the template language too, but that's also not trivial.
- Syntax for how to embed child components in the HTML templating would need some thinking about. Just having them as normal tags wouldn't cut it, as component constructors are often parametrised, etc.

Basically, it's a ton of work for something that is inferior to PS syntax in many ways, without any obvious benefits on the other side aside from aesthetic preference. :wink:

I guess it's true that having HTML syntax might be important for situations where you have a designer or someone not familiar with PS building HTML templates or something, but the amount of work to support it means it'll stay outside the scope of this project at least.

**Comment by ThibaudDauce** [comment]
Thanks a lot for your answers! :-)
